wgsl/index.bs

```diff
@@ -7347,13 +7345,344 @@ See [[#expressions]].
 Statements in a [SHORTNAME] program are executed in control flow order.
 See [[#statements]] and [[#function-calls]].
 
-## Uniformity TODO ## {#uniformity}
+## Uniformity ## {#uniformity}
+
+### Terminology and concepts ### {#uniformity-concepts}
+
+The following definitions are merely informative, trying to give an intuition for what the analysis in the next subsection is computing.
+The analysis is what actually defines these concepts, and when a program is valid or breaks the uniformity rules.
+
+Control flow is said to be uniform at a point in the program if all invocations execute in lockstep at that program point.
+An expression is said to be uniform if it is in uniform control flow, and all invocations compute the same value for it.
+A local variable is said to be uniform, if all invocations have the same value in that variable at every program point where it is live.
+
+### Uniformity analysis overview ### {#uniformity-overview}
+
+In this section we specify an analysis that verifies that functions which are only safe to call in uniform control flow (e.g. barriers and derivatives) are only called in such a context.
+
+<div class="note">Note: This analysis has the following desirable properties:
+      - Sound (meaning that it rejects every program that would break the uniformity requirements of builtins)
+      - Linear time complexity (in the number of tokens in the program)
+      - Refactoring a piece of code into a function cannot make it invalid if it was valid before
+      - If the analysis refuses a program, it provides a straightforward chain of implications that can be used to craft a good error message
+</div>
+
+The analysis analyzes each function in turn, verifying that there is a context where it is safe to call this function. It rejects the program as invalid if there is no such context.
```

Comment:
what semantics does "in turn" carry here? I.e. would my analysis implementation that's running parallel on many threads (e.g. a thread per function) be unexpected by the spec?

Reply:
The actual constraint is listed below: a function must be analyzed before its callers. Apart from that you can use as much parallelism as you want. I've very removed the "in turn".